### PR TITLE
nixos/garage: provide multiple versions to provide an upgrade path for future major releases

### PR DESCRIPTION
**Pull request by RaitoBezarius**
###### Description of changes

Garage is akin to PostgreSQL and similar software in terms of state, it holds precious data and a major upgrade can break stuff it the cluster is not properly upgraded (offline + repair + migrate, etc. depending on release notes).

To maximize data safety, this introduces multiple major versions of Garage in nixpkgs and provide matrix tests on those using the #190181 PR (which this PR depend on).

Once NixOS 22.11 and Garage 0.8 is released, we can move the tip of `pkgs.garage` to `pkgs.garage_0_8` and ensure that state version is kept in check with respect to this, plus, we can provide release notes from Garage directly in our NixOS releases.

Introduced:

- `garage_0_7_3`
- `garage_0_8_0`
- marked garage 0.8 as broken on darwin, see #208366
- marked garage 0.7 as EOL

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-servers/garage.nix [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    systemd.services.garage = {
+      description = "Garage Object Storage (S3 compatible)";
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/garage -c ${configFile} server";
+
+        StateDirectory = mkIf (hasPrefix "/var/lib/garage" cfg.settings.data_dir && hasPrefix "/var/lib/garage" cfg.settings.metadata_dir) "garage";
```
Comment (NickCao) [inline nixos/modules/services/web-servers/garage.nix:74]: Should this be an exact match?

Comment (RaitoBezarius) [reply nixos/modules/services/web-servers/garage.nix:74]: I did not understand but do you mean I should do `cfg.settings.data_dir == "/var/lib/garage"` ?

Comment (NickCao) [reply nixos/modules/services/web-servers/garage.nix:74]: Oops, I was wrong about this, leave it as-is is fine.

Comment (RaitoBezarius) [reply nixos/modules/services/web-servers/garage.nix:74]: BTW, the PR about Garage per se is https://github.com/NixOS/nixpkgs/pull/190181 and this one is more about multiple versions & upgrade paths

**Review by 0x4A6F — COMMENTED** [review]
Overall approval of splitting up packages into major releases.

( `garage -> garage_0_7` ) and `garage_0_8` are building fine.

File: nixos/tests/garage/with-3node-replication.nix [hunk 2]
```diff
@@ -1,50 +1,12 @@
...
-in {
-  name = "garage";
-  meta = {
-    maintainers = with pkgs.lib.maintainers; [ raitobezarius ];
+args@{ mkNode, ... }:
+(import ../make-test-python.nix ({ pkgs, ...} :
+{
+  name = "garage-3node-replication";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ raitobezarius ];
```
Comment (0x4A6F) [inline nixos/tests/garage/with-3node-replication.nix:6]:
```diff
-  meta = with pkgs.lib.maintainers; {
-    maintainers = [ raitobezarius ];
+  meta = {
+    maintainers = with pkgs.lib.maintainers; [ raitobezarius ];
```

File: nixos/modules/services/web-servers/garage.nix [hunk 3]
```diff
@@ -8,7 +8,10 @@ let
   configFile = toml.generate "garage.toml" cfg.settings;
 in
 {
-  meta.maintainers = [ maintainers.raitobezarius ];
+  meta = {
+    doc = ./garage-doc.xml;
+    maintainers = [ maintainers.raitobezarius ];
```
Comment (0x4A6F) [inline nixos/modules/services/web-servers/garage.nix:13]:
```diff
-    maintainers = [ maintainers.raitobezarius ];
+    maintainers = with pkgs.lib.maintainers; [ raitobezarius ];
```

File: nixos/tests/garage/basic.nix [hunk 4]
```diff
@@ -0,0 +1,98 @@
+args@{ mkNode, ... }:
+(import ../make-test-python.nix ({ pkgs, ...} : {
+  name = "garage-basic";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ raitobezarius ];
```
Comment (0x4A6F) [inline nixos/tests/garage/basic.nix:5]:
```diff
-  meta = with pkgs.lib.maintainers; {
-    maintainers = [ raitobezarius ];
+  meta = {
+    maintainers = with pkgs.lib.maintainers; [ raitobezarius ];
```

**Review by 0x4A6F — CHANGES_REQUESTED** [review]
Thanks.

File: nixos/tests/garage/with-3node-replication.nix (hunk 2 above)
Comment (RaitoBezarius) [reply nixos/tests/garage/with-3node-replication.nix]: Done!

File: nixos/modules/services/web-servers/garage.nix (hunk 3 above)
Comment (RaitoBezarius) [reply nixos/modules/services/web-servers/garage.nix]: Done!

File: nixos/tests/garage/basic.nix (hunk 4 above)
Comment (RaitoBezarius) [reply nixos/tests/garage/basic.nix]: Done!

**Comment by RaitoBezarius** [comment]
Would you redo a quick review @0x4A6F ?

**Comment by RaitoBezarius** [comment]
I will use the officially released Garage 0.8 now, friendly ping @0x4A6F @teutat3s for review.

File: nixos/modules/services/web-servers/garage.nix [hunk 5]
```diff
@@ -56,10 +59,12 @@ in
     };
 
     package = mkOption {
-      default = pkgs.garage;
-      defaultText = literalExpression "pkgs.garage";
+      # TODO: when 22.11 is released, put a stateVersion check on Garage 0.7.
+      default = if versionOlder stateVersion "23.05" then pkgs.garage_0_8_0
+                else pkgs.garage_0_7;
```
Comment (0x4A6F) [inline nixos/modules/services/web-servers/garage.nix:64]: 22.11 is released, is this a reoccurring TODO?

Is the default set correctly? Might have a look at the `services.postgresql.package` definition.

Comment (RaitoBezarius) [reply nixos/modules/services/web-servers/garage.nix:64]: Need to verify ; the TODO is wrong indeed!

Comment (RaitoBezarius) [reply nixos/modules/services/web-servers/garage.nix]: Fixed!

**Review by 0x4A6F — APPROVED** [review]
Thanks

**Comment by RaitoBezarius** [comment]
I also need to fix the manual but Docbook is too hard for humans

**Comment by RaitoBezarius** [comment]
I vanquished Docbook, @0x4A6F ; I think once checks pass, we are good to go.

**Review by 0x4A6F — APPROVED** [review]

**Pull request merged**